pkg/operator/operator.go

```diff
@@ -241,76 +262,229 @@ func (k *Kontroller) withLeaderElection() error {
+		return
+	}
 
-	if len(justRebootedNodes) > 0 {
-		glog.Infof("Found %d rebooted nodes, setting annotation %q to false", len(justRebootedNodes), constants.AnnotationOkToReboot)
+	// take some number of the rebootable nodes.
+	// remove any old action annotations and add the reboot-action=pre label.
+	// this starts the pre-reboot actions.
+	err = k.markPreRebootActionable()
+	if err != nil {
+		glog.Errorf("Failed to update rebootable nodes: %v", err)
+		return
 	}
+}
 
-	for _, n := range justRebootedNodes {
-		glog.Infof("Setting 'ok-to-reboot=false' for %q", n.Name)
-		if err := k8sutil.SetNodeAnnotations(k.nc, n.Name, map[string]string{
-			constants.AnnotationOkToReboot: constants.False,
-		}); err != nil {
-			glog.Infof("Failed setting annotation %q on node %q to false: %v", constants.AnnotationOkToReboot, n.Name, err)
+func (k *Kontroller) checkPreRebootActions() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes %v", err)
+	}
+
+	preRebootReq, err := labels.NewRequirement(constants.LabelRebootAction, selection.In, []string{constants.RebootActionPre})
+	if err != nil {
+		return fmt.Errorf("Failed to create filter requirement: %v", err)
+	}
+	preRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, preRebootReq)
+
+	for _, n := range preRebootNodes {
+		res := action.ActionsCompleted(n, k.preRebootActions)
+		if res == action.ActionsSucceeded {
+			glog.Infof("Deleting label %q for %q", constants.LabelRebootAction, n.Name)
+			err := k8sutil.DeleteNodeLabels(k.nc, n.Name, []string{constants.LabelRebootAction})
+			if err != nil {
+				glog.Infof("Failed deleting label %q on node %q: %v", constants.LabelRebootAction, n.Name, err)
+				continue
+			}
+
+			// manually call UpdateNodeRetry so that we can make sure the node still wants to reboot
+			glog.Infof("Marking %q ok to reboot", n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				if !wantsRebootSelector.Matches(fields.Set(node.Annotations)) {
+					glog.Warningf("Node %v no longer wanted to reboot while we were trying to mark it so: %v", node.Name, node.Annotations)
+					return
+				}
+
+				node.Annotations[constants.AnnotationOkToReboot] = constants.True
+			})
+			if err != nil {
+				glog.Infof("Failed to set annotation %q on node %q: %v", constants.AnnotationOkToReboot, n.Name, err)
+			}
 		}
 	}
 
-	nodelist, err = k.nc.List(v1meta.ListOptions{})
+	return nil
+}
+
+func (k *Kontroller) checkPostRebootActions() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
 	if err != nil {
-		glog.Infof("Failed listing nodes: %v", err)
-		return
+		return fmt.Errorf("Failed listing nodes %v", err)
 	}
 
-	// Verify no nodes are still in the process of rebooting to avoid rebooting N > maxRebootingNodes
+	postRebootReq, err := labels.NewRequirement(constants.LabelRebootAction, selection.In, []string{constants.RebootActionPost})
```

Comment:
If all parts of the requirement are constants or known upfront, checkout `MustRequirement` and a var declaration.